src/python/pants/backend/jvm/subsystems/jar_dependency_management.py

```diff
@@ -361,19 +362,24 @@ def _compute_artifact_sets(self, targets):
   def _library_targets(self, managed_jar_dependencies):
     for spec in managed_jar_dependencies.library_specs:
       for target in self.context.resolve(spec):
-        yield target # NB(gmalmquist): I don't think this needs to be transitive.
+        for t in target.closure():
+          yield (t, spec)
 
   def _jar_iterator(self, managed_jar_dependencies):
     for jar in managed_jar_dependencies.payload.artifacts:
       yield jar
-    for dep in self._library_targets(managed_jar_dependencies):
+    for dep, spec in self._library_targets(managed_jar_dependencies):
       if isinstance(dep, JarLibrary):
         for jar in dep.jar_dependencies:
           yield jar
+      elif type(dep) == Target:
+        pass
       else:
         raise TargetDefinitionException(managed_jar_dependencies,
-                                        'Artifacts must be jar() objects or the addresses of '
-                                        'jar_library objects.')
+                                        'Artifacts must be jar() objects, the address of a jar_library(), '
+                                        'or a target() with either of those target types as its transitive '
+                                        'dependencies. Spec {} does not refer to any of '
```

Comment:
Is it the spec of the transitive dependency though? The goal of including the spec in the error message is to identify the original spec in the managed_jar_dependencies target that is in error.